### PR TITLE
HelpActivity converted to kotlin

### LINKED ISSUE #1891: Convert HelpActivity to Kotlin
**Is your feature request related to a problem? Please describe.**
HelpActivity is in java.

**Describe the solution you'd like**
Change it to kotlin.

**Additional context**
Parent issue #1826

### PR DESCRIPTION
**Pull request by yash1ts**
Fixes #1891 
delete HelpActivity.java
added HelpActivity.kt

### CODE REVIEW

File: core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import org.kiwix.kiwixmobile.core.R
+import org.kiwix.kiwixmobile.core.base.BaseActivity
+import org.kiwix.kiwixmobile.core.di.components.CoreComponent
+import org.kiwix.kiwixmobile.core.utils.Constants
+import org.kiwix.kiwixmobile.core.utils.LanguageUtils.Companion.getCurrentLocale
+import java.util.HashMap
+
+class HelpActivity : BaseActivity() {
+  private val titleDescriptionMap =
+    HashMap<String, String>()
```
Comment (abdulwd) [inline core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt]: Can be moved to single line.

File: core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+  private lateinit var toolbar: Toolbar
+
+  private lateinit var recyclerView: RecyclerView
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_help)
+
+    findViewById<TextView>(R.id.activity_help_feedback_text_view)
```
Comment (abdulwd) [inline core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt]: Use Kotlin Android extensions.

**Comment by yash1ts** [comment]
@abdulwd implemented the requested changes. Can you help me with the failing checks, I don't get whats wrong?

**Comment by abdulwd** [comment]
Instead of making a PR from your fork try making a PR from the main repo branch.

**Comment by yash1ts** [comment]
@abdulwd I don't understand what main repo branch?

**Comment by abdulwd** [comment]
I meant https://github.com/kiwix/kiwix-android/

**Comment by yash1ts** [comment]
Created another pull request #1904 from main repo, but still got same error.

File: core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import kotlinx.android.synthetic.main.activity_help.activity_help_feedback_text_view
+import kotlinx.android.synthetic.main.activity_help.activity_help_recycler_view
+import kotlinx.android.synthetic.main.layout_toolbar.toolbar
+
+class HelpActivity : BaseActivity() {
+  private val titleDescriptionMap = HashMap<String, String>()
+
+  private lateinit var mToolbar: Toolbar
+
+  private lateinit var recyclerView: RecyclerView
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt]: get rid of `mToolbar` and `recyclerview`

File: core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    )
+    recyclerView.adapter = HelpAdapter(titleDescriptionMap)
+  }
+
+  private fun sendFeedback() {
+    val intent = Intent(Intent.ACTION_SENDTO)
+    val uriText =
+      "mailto:${Uri.encode(Constants.CONTACT_EMAIL_ADDRESS)}" +
+        "?subject=${Uri.encode("Feedback in ${getCurrentLocale(this).displayLanguage}")}"
+    val uri = Uri.parse(uriText)
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt]: get rid of this variable, `intent.data= Uri.parse(uriText)`

File: core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+import org.kiwix.kiwixmobile.core.utils.Constants
+import org.kiwix.kiwixmobile.core.utils.LanguageUtils.Companion.getCurrentLocale
+import java.util.HashMap
+import kotlinx.android.synthetic.main.activity_help.activity_help_feedback_image_view
+import kotlinx.android.synthetic.main.activity_help.activity_help_feedback_text_view
+import kotlinx.android.synthetic.main.activity_help.activity_help_recycler_view
+import kotlinx.android.synthetic.main.layout_toolbar.toolbar
+
+class HelpActivity : BaseActivity() {
+  private val titleDescriptionMap = HashMap<String, String>()
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt]: I would use `by lazy{ mapOf(R.string.help_2 to R.array.description_help_2, ...).mapKeys{..}.mapValues{ } }`

File: core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_help)
+
+    activity_help_feedback_text_view.setOnClickListener { sendFeedback() }
+    activity_help_feedback_image_view.setOnClickListener { sendFeedback() }
+    setSupportActionBar(toolbar)
+    toolbar.setNavigationOnClickListener { onBackPressed() }
+    if (supportActionBar != null) {
```
Comment (abdulwd) [inline core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt]: This if block can shortened in kotlin.

**Comment by abdulwd** [comment]
Merge master & try running unit tests locally. Check if all of them pass. Then push the changes.

File: core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+import kotlinx.android.synthetic.main.activity_help.activity_help_feedback_image_view
+import kotlinx.android.synthetic.main.activity_help.activity_help_feedback_text_view
+import kotlinx.android.synthetic.main.activity_help.activity_help_recycler_view
+import kotlinx.android.synthetic.main.layout_toolbar.toolbar
+
+class HelpActivity : BaseActivity() {
+  private val titleDescriptionMap by lazy {
+    hashMapOf(R.string.help_2 to R.array.description_help_2,
+              R.string.help_5 to R.array.description_help_5
+    ).entries.associateTo(hashMapOf<String, String>(),
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt]: we can improve the formatting of this a bit, with a sequence of method calls they should go on a new line.

I am a bit surprised ktlint allowed this but the build did pass.

Also I might have been misguided telling you to start with a map, if you start with a `listOf` our existing pairs then you can use `associate` to create our desired map and that will cut down on the operators here.

Also prefer mapOf instead of `hashMapOf`, we generally don't want to leak the underlying implementation of map

File: core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt (hunk 6 above)
Comment (macgills) [reply core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt]: As abdul said we can use `?.let{}` instead of a null check and that can allow us to not use `!!` which is very frequently a code smell to see this operator

File: core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+      DividerItemDecoration(this, DividerItemDecoration.VERTICAL)
+    )
+    activity_help_recycler_view.adapter = HelpAdapter(titleDescriptionMap)
+  }
+
+  private fun sendFeedback() {
+    val intent = Intent(Intent.ACTION_SENDTO)
+    val uriText = "mailto:${Uri.encode(Constants.CONTACT_EMAIL_ADDRESS)}" +
+        "?subject=${Uri.encode("Feedback in ${getCurrentLocale(this).displayLanguage}")}"
+    intent.data = Uri.parse(uriText)
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt]: Actually we have `core-ktx` imported so String has a `toUri()` function if I am not mistaken.

File: core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt (hunk 7 above)
Comment (yash1ts) [reply core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt]: Ok, I will try that.

File: core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt (hunk 8 above)
Comment (yash1ts) [reply core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt]: Yes, you are right.

**Comment by yash1ts** [comment]
> Merge master & try running unit tests locally. Check if all of them pass. Then push the changes.

I tried many times but androidTests are not passing. Actually, I tried running test on previous java code, it failed too.

**Comment by abdulwd** [comment]
Please resolve the conflicts.

**Comment by yash1ts** [comment]
I ran local tests but it fails for verifyHelpActivity(). I don't know what to do. Here is logcat :
[logcat.txt](https://github.com/kiwix/kiwix-android/files/4397487/logcat.txt)

File: core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+import kotlinx.android.synthetic.main.activity_help.activity_help_recycler_view
+import kotlinx.android.synthetic.main.layout_toolbar.toolbar
+import java.util.HashMap
+
+class HelpActivity : BaseActivity() {
+  private val titleDescriptionMap by lazy {
+    listOf(
+      R.string.help_2 to R.array.description_help_2,
+      R.string.help_5 to R.array.description_help_5
+    )
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt]: Sorry, I was unclear in my original comment, keep the associate code up here

File: core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+      it.setTitle(R.string.menu_help)
+    }
+    activity_help_recycler_view.addItemDecoration(
+      DividerItemDecoration(this, DividerItemDecoration.VERTICAL)
+    )
+    activity_help_recycler_view.adapter = HelpAdapter(
+      titleDescriptionMap.associate {
+        getString(it.first) to resources.getStringArray(it.second)
+          .joinToString(separator = System.getProperty("line.separator")!!)
+      } as HashMap<String, String>?
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt]: I don't want to see this cast to a HashMap, yo might have to change the type in HelpAdapter to just Map

Comment (yash1ts) [reply core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt]: Ok

**Comment by yash1ts** [comment]
@macgills I have a question. HelpActivity extends BaseActivity so Butterknife binds its views. But since we are using android extensions here, butterknife may be creating some problem. Actually, app crashes in first-run but clean up project and run again then it works fine.

**Comment by macgills** [comment]
@yash1ts Butterknife only does something if you have annotated a field `@BindView`. Post your exception if you are having trouble, or you can contact me on slack

File: core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt [hunk 11]
```diff
@@ -31,14 +31,16 @@ import kotlinx.android.synthetic.main.activity_help.activity_help_feedback_image
...
 import kotlinx.android.synthetic.main.layout_toolbar.toolbar
-import java.util.HashMap
 
 class HelpActivity : BaseActivity() {
   private val titleDescriptionMap by lazy {
     listOf(
       R.string.help_2 to R.array.description_help_2,
       R.string.help_5 to R.array.description_help_5
-    )
+    ).associate {
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt]: Sorry I keep nitpicking this function.

We can also use a [destructuring declaration](https://kotlinlang.org/docs/reference/multi-declarations.html#destructuring-in-lambdas-since-11) in the associate lambda because `it.first/second` aren't very descriptive

File: core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+import kotlinx.android.synthetic.main.activity_help.activity_help_feedback_text_view
+import kotlinx.android.synthetic.main.activity_help.activity_help_recycler_view
+import kotlinx.android.synthetic.main.layout_toolbar.toolbar
+
+class HelpActivity : BaseActivity() {
+  private val titleDescriptionMap by lazy {
+    listOf(
+      R.string.help_2 to R.array.description_help_2,
+      R.string.help_5 to R.array.description_help_5
+    ).associate { (first, second) ->
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt]: > it.first/second aren't very descriptive

Your kept their names. I want better names than first & second

**Comment by macgills** [comment]
image

This has to read `Fixes #1891` for the merging of this pr to automatically close the linked issue.

You also have to resolve merge conflicts, while you were working on this somebody else made changes to HelpActivity file in java, you will now have to port these features to HelpActivity in kotlin

**Comment by macgills** [comment]
@yash1ts are you having trouble with resolving the conflicts?

image

There are detailed instructions [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/syncing-a-fork) for the commands you need to run

**Comment by macgills** [comment]
HelpActivityTest is now failing because of a subtle change to the composition of the strings

File: core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt [hunk 13]
```diff
@@ -39,7 +39,7 @@ class HelpActivity : BaseActivity() {
       R.string.help_5 to R.array.description_help_5
     ).associate { (title, description) ->
       getString(title) to resources.getStringArray(description)
-        .joinToString(separator = System.getProperty("line.separator")!!)
+        .joinToString(separator = "\n", postfix = "\n")
```
Comment (abdulwd) [inline core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt:42]: Why this change?

Comment (yash1ts) [reply core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt:42]: The HelpActivityTest is written to find view with text like this.

Comment (abdulwd) [reply core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt:42]: Have you tried running the test locally after making this change? Did it pass?

Comment (yash1ts) [reply core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt:42]: Yes I tried but they didn't pass. It couldn't find any views at all.

Comment (macgills) [reply core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpActivity.kt:42]: Revert this and change the test code to use join to string, it putting a `/n` at the end is lazy anyway when joinToString is a much better way
